btf/marshal.go

```diff
@@ -154,23 +127,39 @@ func (e *encoder) Encode() ([]byte, error) {
 		StringLen: uint32(stringLen),
 	}
 
-	err := binary.Write(sliceWriter(buf[:btfHeaderLen]), e.opts.ByteOrder, header)
+	err = binary.Write(sliceWriter(buf[:btfHeaderLen]), e.byteOrder, header)
 	if err != nil {
-		return nil, fmt.Errorf("can't write header: %v", err)
+		return fmt.Errorf("write header: %v", err)
+	}
+
+	return nil
+}
+
+// id returns the ID for the given type or panics with an error.
+func (e *encoder) id(typ Type) TypeID {
+	if _, ok := typ.(*Void); ok {
+		return 0
+	}
+
+	id, ok := e.ids[typ]
+	if !ok {
+		panic(fmt.Errorf("no ID for type %s", typ))
 	}
 
-	return buf, nil
+	return id
 }
 
-func (e *encoder) deflateType(typ Type) (err error) {
-	raw := &e.raw
+func (e *encoder) deflateType(typ Type, raw *rawType) (err error) {
```

Comment:
This is really awkward and counter-intuitive. Assuming escaping is the problem, can't we keep `encoder.raw` for scratch space? Maybe the real answer is rejigging/removing `rawType`, but I'm not sure what the problem is here.

Reply:
Fixed